### PR TITLE
fix: Improve handling of unknown hosts and missing certs

### PR DESCRIPTION
**Pull request by rhansen**
* fix: Don't create cert error https server if https is not enabled
  * fix: Emit TLS error if there are no certs available
  * fix: Don't create fallback http(s) server when http(s) disabled

plus a couple of cleanups

### CODE REVIEW

File: nginx.tmpl [hunk 1]
```diff
@@ -454,22 +462,71 @@ server {
...
+                 *
+                 * Firefox:
+                 * > Secure Connection Failed
+                 * >
+                 * > An error occurred during a connection to example.test.
+                 * > Peer reports it experienced an internal error.
+                 * >
+                 * > Error code: SSL_ERROR_INTERNAL_ERROR_ALERT "TLS error".
+                 *
+                 */}}
```
Comment (buchdag) [inline nginx.tmpl:552]: Could you add this to the documentation ?

**Review by buchdag — CHANGES_REQUESTED** [review]

File: nginx.tmpl (hunk 1 above)
Comment (rhansen) [reply nginx.tmpl]: Done.

File: README.md [hunk 2]
```diff
@@ -347,17 +347,41 @@ Note that the `Mozilla-Old` policy should use a 1024 bits DH key for compatibili
...
 
 By default, [HTTP Strict Transport Security (HSTS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security)  is enabled with `max-age=31536000` for HTTPS sites. You can disable HSTS with the environment variable `HSTS=off` or use a custom HSTS configuration like `HSTS=max-age=31536000; includeSubDomains; preload`. 
 
 *WARNING*: HSTS will force your users to visit the HTTPS version of your site for the `max-age` time - even if they type in `http://` manually.  The only way to get to an HTTP site after receiving an HSTS response is to clear your browser's HSTS cache.
 
+#### Missing Certificate
+
+If HTTPS is enabled for a virtual host but its certificate is missing, nginx-proxy will configure nginx to use the default certificate (`default.crt` with `default.key`) and return a 500 error.
+
+If the default certificate is also missing, nginx-proxy will configure nginx to accept HTTPS connections but fail the TLS negotiation.  Client browsers will will render a TLS error page.  As of March 2023, web browsers display the following error messages:
```
Comment (buchdag) [inline README.md:364]: typo:
```suggestion
If the default certificate is also missing, nginx-proxy will configure nginx to accept HTTPS connections but fail the TLS negotiation.  Client browsers will render a TLS error page.  As of March 2023, web browsers display the following error messages:
```

**Review by buchdag — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (rhansen) [reply README.md]: Done.

**Review by buchdag — DISMISSED** [review]
Impressive work, thank you.

**Comment by buchdag** [comment]
@rhansen as this changes nginx.tmpl considerably, I'm not certain it won't interfere with #1934 and #2173, could you rebase this branch on `main` ?

**Review by buchdag — APPROVED** [review]

**Comment by rhansen** [comment]
Thank you for rebasing and merging, @buchdag!

**Comment by SchoNie** [comment]
@rhansen 
Sorry for bumping this PR. But there seems to be a problem with Secure Renegotiation (RFC 5746) after this merge Secure Renegotiation is not working anymore.
So instead of using the SSL cache, clients have to start the TLS handshake again each request.

I am testing this with `docker run --rm -ti drwetter/testssl.sh url.com` which reports: **Secure Renegotiation (RFC 5746)          OpenSSL handshake didn't succeed**

And SSLLabs reports: Session resumption (caching): **No (IDs assigned but not accepted)**

Or testing with openssl directly: `openssl s_client -connect hostname:443 -reconnect -CAfile cacert.pem`
Before this PR i am assigned a "Post-Handshake New Session Ticket arrived", after this PR no ticket.

I commented [line 415 in the nginx.tmpl](https://github.com/nginx-proxy/nginx-proxy/blob/main/nginx.tmpl#L415) and that fixed it for me.
I am not sure what is happening, but could it be the fallback listener takes precedence in the openssl handshake?

**Comment by buchdag** [comment]
@SchoNie line 415 is a listener on the HTTP port, I don't get how it could influence something TLS related 🤯

Could you show us the rendered nginx configuration that work and the one that doesn't (you can get it by executing `nginx -T` inside the nginx-proxy container) ?

**Comment by SchoNie** [comment]
Ofcourse, sorry for not doing that right away. I created a small whoami (my prod setup is huge) setup and was able to reproduce.
They are in the collapsed block to not spam this with huge code blocks.

Let me know if any more is needed.

<details>
  <summary>Failing nginx.conf:</summary>

```
# nginx-proxy version : 1.3.0-17-g7dd32e3
# Networks available to the container running docker-gen (which are assumed to
# match the networks available to the container running nginx):
#     test_default
# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
# scheme used to connect to this server
map $http_x_forwarded_proto $proxy_x_forwarded_proto {
    default $scheme;
    '' $scheme;
}
map $http_x_forwarded_host $proxy_x_forwarded_host {
    default $http_host;
    '' $http_host;
}
# If we receive X-Forwarded-Port, pass it through; otherwise, pass along the
# server port the client connected to
map $http_x_forwarded_port $proxy_x_forwarded_port {
    default $server_port;
    '' $server_port;
}
# If the request from the downstream client has an "Upgrade:" header (set to any
# non-empty value), pass "Connection: upgrade" to the upstream (backend) server.
# Otherwise, the value for the "Connection" header depends on whether the user
# has enabled keepalive to the upstream server.
map $http_upgrade $proxy_connection {
    default upgrade;
    '' $proxy_connection_noupgrade;
}
map $upstream_keepalive $proxy_connection_noupgrade {
    # Preserve nginx's default behavior (send "Connection: close").
    default close;
    # Use an empty string to cancel nginx's default behavior.
    true '';
}
# Abuse the map directive (see <https://stackoverflow.com/q/14433309>) to ensure
# that $upstream_keepalive is always defined.  This is necessary because:
#   - The $proxy_connection variable is indirectly derived from
#     $upstream_keepalive, so $upstream_keepalive must be defined whenever
#     $proxy_connection is resolved.
#   - The $proxy_connection variable is used in a proxy_set_header directive in
#     the http block, so it is always fully resolved for every request -- even
#     those where proxy_pass is not used (e.g., unknown virtual host).
map "" $upstream_keepalive {
    # The value here should not matter because it should always be overridden in
    # a location block (see the "location" template) for all requests where the
    # value actually matters.
    default false;
}
# Apply fix for very long server names
server_names_hash_bucket_size 128;
# Default dhparam
ssl_dhparam /etc/nginx/dhparam/dhparam.pem;
# Set appropriate X-Forwarded-Ssl header based on $proxy_x_forwarded_proto
map $proxy_x_forwarded_proto $proxy_x_forwarded_ssl {
    default off;
    https on;
}
gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
log_format vhost '$host $remote_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent "$http_referer" "$http_user_agent" "$upstream_addr" "$request_time" $ssl_protocol/$ssl_cipher';
access_log off;
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_ciphers 'ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384';
    ssl_prefer_server_ciphers off;
error_log /dev/stderr;
resolver 127.0.0.11;
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Host $proxy_x_forwarded_host;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;
proxy_set_header X-Original-URI $request_uri;
# Mitigate httpoxy attack (see README for details)
proxy_set_header Proxy "";
server {
    server_name _; # This is just an invalid value which will never trigger on a real hostname.
    server_tokens off;
    listen 80;
    listen 443 ssl http2;
    access_log /var/log/nginx/access.log vhost;
    # No default.crt certificate found for this vhost, so force nginx to emit a
    # TLS error if the client connects via https.
    ssl_ciphers aNULL;
    set $empty "";
    ssl_certificate data:$empty;
    ssl_certificate_key data:$empty;
    if ($https) {
        return 444;
    }
    return 503;
}
# test.domain.com/
upstream test.domain.com-42099b4af021e53fd8fd4e056c2568d7c2e3ffa8 {
    # Container: test-whoami-1
    #     networks:
    #         test_default (reachable)
    #     IP address: 172.22.0.2
    #     exposed ports: 80/tcp
    #     default port: 80
    #     using port: 8000
    server 172.22.0.2:8000;
    # Container: test-whoami-1
    #     networks:
    #         test_default (reachable)
    #     IP address: 172.22.0.2
    #     exposed ports: 80/tcp
    #     default port: 80
    #     using port: 8000
    server 172.22.0.2:8000;
    keepalive 32;
}
server {
    server_name test.domain.com;
    listen 80 ;
    access_log /var/log/nginx/access.log vhost;
    # Do not HTTPS redirect Let's Encrypt ACME challenge
    location ^~ /.well-known/acme-challenge/ {
        auth_basic off;
        auth_request off;
        allow all;
        root /usr/share/nginx/html;
        try_files $uri =404;
        break;
    }
    location / {
        return 301 https://$host$request_uri;
    }
}
server {
    server_name test.domain.com;
    access_log /var/log/nginx/access.log vhost;
    listen 443 ssl http2 ;
    ssl_session_timeout 5m;
    ssl_session_cache shared:SSL:50m;
    ssl_session_tickets off;
    ssl_certificate /etc/nginx/certs/domain.com.crt;
    ssl_certificate_key /etc/nginx/certs/domain.com.key;
    ssl_stapling on;
    ssl_stapling_verify on;
    ssl_trusted_certificate /etc/nginx/certs/domain.com.chain.pem;
    set $sts_header "";
    if ($https) {
        set $sts_header "max-age=31536000; includeSubDomains; preload";
    }
    add_header Strict-Transport-Security $sts_header always;
    location / {
        proxy_pass http://test.domain.com-42099b4af021e53fd8fd4e056c2568d7c2e3ffa8;
        set $upstream_keepalive true;
    }
}
```
</details>


Then I edit the first listener, and do a `nginx -s reload`
<details>
  <summary>Working nginx.conf:</summary>

```
# nginx-proxy version : 1.3.0-17-g7dd32e3
# Networks available to the container running docker-gen (which are assumed to
# match the networks available to the container running nginx):
#     test_default
# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
# scheme used to connect to this server
map $http_x_forwarded_proto $proxy_x_forwarded_proto {
    default $scheme;
    '' $scheme;
}
map $http_x_forwarded_host $proxy_x_forwarded_host {
    default $http_host;
    '' $http_host;
}
# If we receive X-Forwarded-Port, pass it through; otherwise, pass along the
# server port the client connected to
map $http_x_forwarded_port $proxy_x_forwarded_port {
    default $server_port;
    '' $server_port;
}
# If the request from the downstream client has an "Upgrade:" header (set to any
# non-empty value), pass "Connection: upgrade" to the upstream (backend) server.
# Otherwise, the value for the "Connection" header depends on whether the user
# has enabled keepalive to the upstream server.
map $http_upgrade $proxy_connection {
    default upgrade;
    '' $proxy_connection_noupgrade;
}
map $upstream_keepalive $proxy_connection_noupgrade {
    # Preserve nginx's default behavior (send "Connection: close").
    default close;
    # Use an empty string to cancel nginx's default behavior.
    true '';
}
# Abuse the map directive (see <https://stackoverflow.com/q/14433309>) to ensure
# that $upstream_keepalive is always defined.  This is necessary because:
#   - The $proxy_connection variable is indirectly derived from
#     $upstream_keepalive, so $upstream_keepalive must be defined whenever
#     $proxy_connection is resolved.
#   - The $proxy_connection variable is used in a proxy_set_header directive in
#     the http block, so it is always fully resolved for every request -- even
#     those where proxy_pass is not used (e.g., unknown virtual host).
map "" $upstream_keepalive {
    # The value here should not matter because it should always be overridden in
    # a location block (see the "location" template) for all requests where the
    # value actually matters.
    default false;
}
# Apply fix for very long server names
server_names_hash_bucket_size 128;
# Default dhparam
ssl_dhparam /etc/nginx/dhparam/dhparam.pem;
# Set appropriate X-Forwarded-Ssl header based on $proxy_x_forwarded_proto
map $proxy_x_forwarded_proto $proxy_x_forwarded_ssl {
    default off;
    https on;
}
gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
log_format vhost '$host $remote_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent "$http_referer" "$http_user_agent" "$upstream_addr" "$request_time" $ssl_protocol/$ssl_cipher';
access_log off;
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_ciphers 'ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384';
    ssl_prefer_server_ciphers off;
error_log /dev/stderr;
resolver 127.0.0.11;
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Host $proxy_x_forwarded_host;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;
proxy_set_header X-Original-URI $request_uri;
# Mitigate httpoxy attack (see README for details)
proxy_set_header Proxy "";
server {
    server_name _; # This is just an invalid value which will never trigger on a real hostname.
    server_tokens off;
    listen 80;
**#    listen 443 ssl http2; <<this line is commented**
    access_log /var/log/nginx/access.log vhost;
    # No default.crt certificate found for this vhost, so force nginx to emit a
    # TLS error if the client connects via https.
    ssl_ciphers aNULL;
    set $empty "";
    ssl_certificate data:$empty;
    ssl_certificate_key data:$empty;
    if ($https) {
        return 444;
    }
    return 503;
}
# test.domain.com/
upstream test.domain.com-42099b4af021e53fd8fd4e056c2568d7c2e3ffa8 {
    # Container: test-whoami-1
    #     networks:
    #         test_default (reachable)
    #     IP address: 172.22.0.2
    #     exposed ports: 80/tcp
    #     default port: 80
    #     using port: 8000
    server 172.22.0.2:8000;
    # Container: test-whoami-1
    #     networks:
    #         test_default (reachable)
    #     IP address: 172.22.0.2
    #     exposed ports: 80/tcp
    #     default port: 80
    #     using port: 8000
    server 172.22.0.2:8000;
    keepalive 32;
}
server {
    server_name test.domain.com;
    listen 80 ;
    access_log /var/log/nginx/access.log vhost;
    # Do not HTTPS redirect Let's Encrypt ACME challenge
    location ^~ /.well-known/acme-challenge/ {
        auth_basic off;
        auth_request off;
        allow all;
        root /usr/share/nginx/html;
        try_files $uri =404;
        break;
    }
    location / {
        return 301 https://$host$request_uri;
    }
}
server {
    server_name test.domain.com;
    access_log /var/log/nginx/access.log vhost;
    listen 443 ssl http2 ;
    ssl_session_timeout 5m;
    ssl_session_cache shared:SSL:50m;
    ssl_session_tickets off;
    ssl_certificate /etc/nginx/certs/domain.com.crt;
    ssl_certificate_key /etc/nginx/certs/domain.com.key;
    ssl_stapling on;
    ssl_stapling_verify on;
    ssl_trusted_certificate /etc/nginx/certs/domain.com.chain.pem;
    set $sts_header "";
    if ($https) {
        set $sts_header "max-age=31536000; includeSubDomains; preload";
    }
    add_header Strict-Transport-Security $sts_header always;
    location / {
        proxy_pass http://test.domain.com-42099b4af021e53fd8fd4e056c2568d7c2e3ffa8;
        set $upstream_keepalive true;
    }
}
```
</details>

<details>
  <summary>The docker-compose.yml I use for testing:</summary>
```
version: '3.9'

services:
  nginx-proxy:
    image: nginxproxy/nginx-proxy:alpine
    container_name: nginx-proxy
    restart: unless-stopped
    environment:
      - TZ=Europe/Amsterdam
      - SSL_POLICY=Mozilla-Intermediate
      - HSTS=max-age=31536000; includeSubDomains; preload
      - TRUST_DOWNSTREAM_PROXY=false
      - LOG_FORMAT=$$host $$remote_addr - $$remote_user [$$time_local] "$$request" $$status $$body_bytes_sent "$$http_referer" "$$http_user_agent" "$$upstream_addr" "$$request_time" $$ssl_protocol/$$ssl_cipher
    ports:
      - 80:80
      - 443:443
    volumes:
      - "/var/run/docker.sock:/tmp/docker.sock:ro"
      - "./nginx/conf.d/:/etc/nginx/conf.d/"      
      - "./nginx/vhost.d/:/etc/nginx/vhost.d/"
      - "./nginx/certs/:/etc/nginx/certs/"
      - "./accesslogs/:/var/log/nginx/"
      - "./templates/nginx_custom.tmpl:/app/nginx.tmpl:ro"

  whoami:
    image: traefik/whoami:latest
    restart: unless-stopped
    environment:
      - WHOAMI_PORT_NUMBER=8000
      - VIRTUAL_HOST=test.domain.com
      - VIRTUAL_PORT=8000
      - VIRTUAL_PATH=/
    labels:
      com.github.nginx-proxy.nginx-proxy.keepalive: "32"
```
</details>

**Comment by buchdag** [comment]
@SchoNie you are right about the cause but that should not happen if the tests correctly handle SNI, I'm beginning to suspect that those tests don't and aren't reflecting the reality.

Could you try again with:
```console
openssl s_client -connect test.domain.com:443 -servername test.domain.com -reconnect -CAfile cacert.pem
```

If you don't use `-servername test.domain.com` pretty much every tests based on `openssl s_client` will fail on an SNI based reverse proxy like nginx-proxy.

https://major.io/2012/02/07/using-openssls-s_client-command-with-web-servers-using-server-name-indication-sni/

**Comment by terryzwt** [comment]
After using this feature, got below error.
```
2023/06/03 08:12:11 [warn] 252#252: *902 using uninitialized "empty" variable while SSL handshaking, client: 1.1.1.1, server: 0.0.0.0:443
2023/06/03 08:12:11 [error] 252#252: *902 cannot load certificate "data:": PEM_read_bio_X509_AUX() failed (SSL: error:0480006C:PEM routines::no start line:Expecting: TRUSTED CERTIFICATE) while SSL handshaking, client: 1.1.1.1, server: 0.0.0.0:443
```

in my case, the /etc/nginx/certs/default.key & default.crt is missing, and the generated conf like this:
```
server {
    server_name _; # This is just an invalid value which will never trigger on a real hostname.
    server_tokens off;
    listen 80;
    listen 443 ssl http2;
    ssl_session_cache shared:SSL:50m;
    ssl_session_tickets off;
    # No default.crt certificate found for this vhost, so force nginx to emit a
    # TLS error if the client connects via https.
    ssl_ciphers aNULL;
    set $empty "";
    ssl_certificate data:$empty;
    ssl_certificate_key data:$empty;
    if ($https) {
        return 444;
    }
    return 503;
}
```

**Comment by buchdag** [comment]
@terryzwt what feature are you talking about ? (this PR was more about fixes and refactor)

**Comment by terryzwt** [comment]
@buchdag I mean this fixed cause error message like below, it should be easy to reproduce.
```
2023/06/03 08:12:11 [warn] 252#252: *902 using uninitialized "empty" variable while SSL handshaking, client: 1.1.1.1, server: 0.0.0.0:443
2023/06/03 08:12:11 [error] 252#252: *902 cannot load certificate "data:": PEM_read_bio_X509_AUX() failed (SSL: error:0480006C:PEM routines::no start line:Expecting: TRUSTED CERTIFICATE) while SSL handshaking, client: 1.1.1.1, server: 0.0.0.0:443
```

**Comment by vincent-herlemont** [comment]
From what I understand in this test, with this fix it is no longer possible with a self-signed certificate to access a site with an unknown domain name. So, if ever we have an architecture that needs this, it will no longer work.


https://github.com/rhansen/nginx-proxy/blob/9b4bb07b348dc5a428b94416517291adb30794c3/test/test_fallback.py#L39-L92
```
("withdefault.yml", "http://unknown.nginx-proxy.test/", 503, None),
("withdefault.yml", "https://unknown.nginx-proxy.test/", 503, None),
...
("nodefault.yml", "http://unknown.nginx-proxy.test/", 503, None),
("nodefault.yml", "https://unknown.nginx-proxy.test/", None, INTERNAL_ERR_RE),
...
("nohttp.yml", "http://unknown.nginx-proxy.test/", None, CONNECTION_REFUSED_RE),
("nohttp.yml", "https://unknown.nginx-proxy.test/", 503, None),
...
("nohttp-on-app.yml", "http://unknown.nginx-proxy.test/", None, CONNECTION_REFUSED_RE),
("nohttp-on-app.yml", "https://unknown.nginx-proxy.test/", 503, None),
...
("nohttp-with-missing-cert.yml", "http://unknown.nginx-proxy.test/", 503, None),
("nohttp-with-missing-cert.yml", "https://unknown.nginx-proxy.test/", 503, None),
...
("nohttps.yml", "http://unknown.nginx-proxy.test/", 503, None),
("nohttps.yml", "https://unknown.nginx-proxy.test/", None, CONNECTION_REFUSED_RE),
...
("nohttps-on-app.yml", "http://unknown.nginx-proxy.test/", 503, None),
("nohttps-on-app.yml", "https://unknown.nginx-proxy.test/", None, CONNECTION_REFUSED_RE),
```

Two things, it seems to me that consequently
- this is not a fix but a change in behavior.
- find a solution or an explanation for this behavior that I may not have understood.

I will open an issue  for this, as it seems I have the same error as @terryzwt .

**Comment by buchdag** [comment]
Hi.

Provided I understood correctly the issue you're facing: accessing a service with an unknow domain name was unfortunately never and advertised nor supported feature, it only worked as a side effect of the proxy incorrectly redirecting request for unknown hostname to a proxied service, which was a real issue for a long time.

I believe your use case (again if I understood it correctly) can still work by using `DEFAULT_HOST` and providing a default certificate and private key.

**Pull request merged**